modules/in_band_dtmf/in_band_dtmf.c

```diff
@@ -0,0 +1,323 @@
+/**
+ * @file in_band_dtmf.c  DTMF decoder
+ */
+#include <time.h>
+#include <ctype.h>
+#include <string.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+
+
+/**
+ * @defgroup in_band_dtmf in_band_dtmf
+ *
+ * Audio filter that de- and encodes DTMF tones.
+ *
+ * The tone length of an encoded tone is 0.1s.
+ * New tones can be added while encoding is active.
+ */
+
+struct in_band_dtmf_filt_dec {
+	struct aufilt_dec_st af;  /* inheritance */
+	struct dtmf_dec *dec;
+	const struct audio *au;
+	struct le le_priv;
+	struct tmr tmr_dtmf_end;
+};
+
+
+struct in_band_dtmf_filt_enc {
+	struct aufilt_enc_st af;  /* inheritance */
+	struct mbuf *mb;
+	unsigned srate;
+	struct le le_priv;
+};
+
+
+static struct list encs;
+static struct list decs;
+
+
+static void dtmfend_handler(void *arg)
+{
+	char digit = (char)(uintptr_t)arg;
+	struct in_band_dtmf_filt_dec *st;
+
+	if (list_isempty(&decs)) {
+		warning("in_band_dtmf: no active call\n");
+		return;
+	}
+
+	st = decs.head->data;
+
+	audio_call_telev_handler(st->au, digit, true);
+}
+
+
+static void in_band_dtmf_dec_handler(char digit, void *arg)
+{
+	(void)arg;
+	struct in_band_dtmf_filt_dec *st;
+
+	if (list_isempty(&decs)) {
+		warning("in_band_dtmf: no active call\n");
+		return;
+	}
+
+	st = decs.head->data;
+
+	tmr_start(&st->tmr_dtmf_end, 50,
+		dtmfend_handler, (void*)(uintptr_t)digit);
+	audio_call_telev_handler(st->au, digit, false);
+}
+
+
+static void enc_destructor(void *arg)
+{
+	struct in_band_dtmf_filt_enc *st =
+			(struct in_band_dtmf_filt_enc *) arg;
+
+	list_unlink(&st->af.le);
+	list_unlink(&st->le_priv);
+	mem_deref(st->mb);
+}
+
+
+static void dec_destructor(void *arg)
+{
+	struct in_band_dtmf_filt_dec *st =
+			(struct in_band_dtmf_filt_dec *) arg;
+
+	list_unlink(&st->af.le);
+	list_unlink(&st->le_priv);
+	tmr_cancel(&st->tmr_dtmf_end);
+	mem_deref(st->dec);
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct in_band_dtmf_filt_enc *st;
+	(void)ctx;
+	(void)af;
+	(void)au;
+
+	if (!stp || !prm)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->mb = mbuf_alloc(0);
+	if (!st->mb) {
```

Comment:
here is a memory leak. Declare a variable `int err`, if mbuf_alloc fails set err to ENOMEM and goto out.

Reply:
fixed by 1b0cf18c